3주차 미션 / 서버 1조 주민석

emes-g opened the pull request:
안녕하세요. 서버 1조 주민석입니다.
스터디 시점까지 미션을 많이 진행하지 못해 아쉬웠는데, 이렇게 끝까지 완수하여 제출할 수 있어 만족했던 3주차였던 것 같습니다.
3주차 미션을 진행하면서 고민했던 부분이나 신경써서 구현했던 부분에 대해 같이 이야기해보고자 합니다.

## 3주차 회고
2주차에 HttpRequest (미션 2-2) 까지 리팩토링하고 PR을 보냈어서, 이번주 월요일까지 이후 리팩토링을 진행하고 3주차 미션을 진행하였습니다. 확실히 2주차 리팩토링을 끝까지 진행해서인지, 3주차 리팩토링을 진행하는 것이 수월한 편이라고 생각하였습니다. 아마 이번 미션만큼 직접적일지는 모르겠으나, 앞으로도 이전 주차의 학습 내용이 스노우볼처럼 영향을 미칠 것이라고 생각해서, 해당 주차가 지나더라도 이해 못한 부분은 더 찾아보고 질문해야겠다는 생각이 들었습니다. 
이후로도 열심히 노력해서 방학 프로젝트까지 완수할 수 있도록 하겠습니다 😁

## 구현 및 리팩토링
2주차 리팩토링에서 RequestHandler-RequestMapper-Controller로 이어지는 구조가, 
3주차 리팩토링에서 DispatchServlet-RequestMapper-Controller로 그대로 이어지는 것을 체감할 수 있었습니다.
최대한 스스로 구현 및 리팩토링을 해보고 정답 코드와 비교한 결과, 다음을 추가적으로 리팩토링 할 수 있었습니다.

### 1. .jsp로 url이 끝나지 않도록 리팩토링
- 특별한 로직 없이 뷰(JSP)에 대한 이동만을 담당하는 ForwardController 생성
- 특정 버튼 클릭 시 home.jsp로 가는 등, url이 jsp로 끝나지 않도록 리팩토링

### 2. DispatcherServlet과 RequestMapper에 명확한 책임을 부여
DistpacherServlet
- 모든 요청을 가장 먼저 받아, 적합한 컨트롤러에 위임
- 비즈니스 로직 처리 결과를 해석하여 forward or redirect 수행

RequestMapper
- forward or redirect는 DispatcherServlet에서 수행하게 함
- 온전히, URL과 컨트롤러 매핑하는 책임에 집중

emes-g committed:
Feat: HomeController 구현 (1-1)

emes-g committed:
Feat: 중복된 head, nav 태그 제거 (1-2)

emes-g committed:
Fix: 중복 태그 제거 중, 파일명 변경 누락 (1-2)

emes-g committed:
Feat: 개인정보 수정 구현 (1-3)

emes-g committed:
Refactor: div 태그 중 nav 관련 부분 중복 제거 (1-2)

emes-g committed:
Feat: 로그인/로그아웃 구현 (1-4)

emes-g committed:
Feat: 회원 목록 및 개인정보 수정 보안 강화 (1-5)

emes-g committed:
Refactor: MVC 1단계 적용 (미션 2)

정답 코드와 힌트를 참고하지 않고, 개인적으로 리팩토링한 버전입니다.
다음 커밋에서는 정답 코드와 비교하여 일부 수정한 버전을 올릴 계획입니다.

emes-g committed:
Refactor: .jsp로 url이 끝나지 않도록 리팩토링

1. 특별한 로직 없이 뷰(JSP)에 대한 이동만을 담당하는 ForwardController 생성
2. 특정 버튼 클릭 시 home.jsp로 가는 등, url이 jsp로 끝나지 않도록
   리팩토링

emes-g committed:
Refactor: DispatcherServlet과 RequestMapper에 명확한 책임을 부여

1. DistpacherServlet
- 모든 요청을 가장 먼저 받아, 적합한 컨트롤러에 위임
- 비즈니스 로직 처리 결과를 해석하여 forward or redirect 수행

2. RequestMapper
- forward or redirect는 DispatcherServlet에서 수행하게 함
- 온전히, URL과 컨트롤러 매핑하는 책임에 집중

kwiyoon reviewed (commented):
안녕하세요! 미션하시느라 수고 정말 많으셨습니다 😁🤗

pr 내용이 딱 제가 원하는 구성인 것 같아요. 물론 제가 절대로 정답은 아니지만, 코드 리뷰어의 입장에서 보고싶은 내용들이 딱 들어가있어요👍👍 커밋도 세세하고 명확하게 작성해주셔서 보기 좋습니다 ㅎㅎ

미션 아주 잘 수행해주셨는데, 몇가지 리뷰 드린 부분들이 있어서 확인해보시면 좋을 거 같아요! 수고 많으셨습니다~~😀

emes-g committed:
Refactor: 3주차 피드백 내용 반영

1. DispatcherServlet에 loadOnStartUp 속성 추가
2. 로그인 확인을 담당하는 유틸 클래스 추가
3. LoginController에서 유저의 비밀번호 검사 과정 수정
- User.matchPassword() 사용

emes-g commented on src/main/java/core/mvc/DispatcherServlet.java:
```diff
@@ -0,0 +1,38 @@
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@WebServlet("/")
```
서블릿 컨테이너(톰캣)를 실행하면서 서블릿을 미리 실행두는 역할인 것 같습니다!
감사합니다.

emes-g commented on src/main/java/jwp/controller/ListUserController.java:
```diff
@@ -1,21 +1,31 @@
+        Object value = session.getAttribute("user");
+
+        // 로그인 상태라면
+        if (value != null) {
```
세션에서 유저와 관련된 정보를 처리하는 클래스를 하나 만들었습니다!
감사합니다.

emes-g commented on src/main/java/jwp/controller/LoginController.java:
```diff
@@ -0,0 +1,29 @@
+    public String execute(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        User user = MemoryUserRepository.getInstance().findUserById(req.getParameter("userId"));
+
+        if (user != null && user.getPassword().equals(req.getParameter("password"))) {
```
User.matchPassword()를 이용하여 리팩토링 해보았습니다!
감사합니다 😀

emes-g committed:
Refactor: add move function in DispatcherServlet

viewName에 해당하는 view를 클라이언트에게 응답하는 함수인 move() 추가

emes-g committed:
Feat: MemoryQuestionRepository에 존재하는 질문 출력

HomeController
- MemoryQuestionRepository에 저장된 질문들 request에 담기

ShowQuestionController
- 질문 번호에 맞는 질문 출력하기

emes-g committed:
Feat: 질문 추가 기능 구현 (CreateQuestionController)

emes-g committed:
Feat: MemoryAnswerRepository에 존재하는 답변(댓글) 출력

ShowQuestionController
- 질문 번호에 맞는 답변들 출력하기
